src/com/amazon/corretto/crypto/provider/EvpKey.java

```diff
@@ -204,4 +208,23 @@ protected static class InternalKey extends NativeResource {
     protected interface CanDerivePublicKey<T extends EvpKey & PublicKey> {
         T getPublicKey();
     }
+
+    // The Key interface requires us to be serializable.
+    // However, we are not easily serializable because we have native state which won't be captured.
+    // So, we prevent any attempt to serialize ourselves by throwing an exception.
+    // If we discover that we actually need to support this functionality, we can manage it through serializing an
+    // encoded copy of the key and decoding that upon deserialization.
+    // Doing that will require changing out InternalKey from being a final field to a transient one and this feels
+    // like a bad tradeoff if we can avoid it.
+    private void writeObject(final ObjectOutputStream out) throws IOException {
```

Comment:
looks like the new tests are failing:

```
  JUnit Jupiter:EvpKeyFactoryTest:cannotSerializeKeys(KeyPair, String):RSA
    MethodSource [className = 'com.amazon.corretto.crypto.provider.test.EvpKeyFactoryTest', methodName = 'cannotSerializeKeys', methodParameterTypes = 'java.security.KeyPair, java.lang.String'
    => org.opentest4j.AssertionFailedError: Expected java.io.NotSerializableException to be thrown, but nothing was thrown.
       org.junit.jupiter.api.AssertThrows.assertThrows(AssertThrows.java:71)
       org.junit.jupiter.api.AssertThrows.assertThrows(AssertThrows.java:37)
       org.junit.jupiter.api.Assertions.assertThrows(Assertions.java:2952)
       com.amazon.corretto.crypto.provider.test.EvpKeyFactoryTest.cannotSerializeKeys(EvpKeyFactoryTest.java:370)

```

i'm guessing that because these methods are `private`, they only apply to this abstract class and aren't inherited by subclasses. going by [this javadoc](https://docs.oracle.com/javase/7/docs/api/java/io/ObjectOutputStream.html), it looks like each individual class implementing this exception throwing behavior needs to override these 3 methods:

> Classes that require special handling during the serialization and deserialization process must implement special methods with these exact signatures:
>
>  ```
>  private void readObject(java.io.ObjectInputStream stream)
>      throws IOException, ClassNotFoundException;
>  private void writeObject(java.io.ObjectOutputStream stream)
>      throws IOException
>  private void readObjectNoData()
>      throws ObjectStreamException;
> ```

perhaps you can declare these abstract on the base class?

Reply:
maybe you can work around this by implementing `Externalizable` on the base class and throwing in each of its 2 methods?

> Implementing the Externalizable interface allows the object to assume complete control over the contents and format of the object's serialized form. The methods of the Externalizable interface, writeExternal and readExternal, are called to save and restore the objects state. When implemented by a class they can write and read their own state using all of the methods of ObjectOutput and ObjectInput. It is the responsibility of the objects to handle any versioning that occurs.

Reply:
I'm guessing this is because I'm accidentally am serializing non-ACCP objects. I'll try to fix this.